src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1444,37 +1444,57 @@ public static unsafe SocketError Select(IList checkRead, IList checkWrite, IList
+                AddToPollArray(events, eventsLength, checkWrite, ref offset, Interop.Sys.PollEvents.POLLOUT, ref refsAdded);
+                AddToPollArray(events, eventsLength, checkError, ref offset, Interop.Sys.PollEvents.POLLPRI, ref refsAdded);
+                Debug.Assert(offset == eventsLength, $"Invalid adds. offset={offset}, eventsLength={eventsLength}.");
+                Debug.Assert(refsAdded == eventsLength, $"Invalid ref adds. refsAdded={refsAdded}, eventsLength={eventsLength}.");
+
+                // Do the poll
+                uint triggered = 0;
+                int milliseconds = microseconds == -1 ? -1 : microseconds / 1000;
+                Interop.Error err = Interop.Sys.Poll(events, (uint)eventsLength, milliseconds, &triggered);
+                if (err != Interop.Error.SUCCESS)
+                {
+                    return GetSocketErrorForErrorCode(err);
+                }
 
-            // Remove from the lists any entries which weren't set
-            if (triggered == 0)
-            {
-                checkRead?.Clear();
-                checkWrite?.Clear();
-                checkError?.Clear();
+                // Remove from the lists any entries which weren't set
+                if (triggered == 0)
+                {
+                    Socket.SocketListDangerousReleaseRefs(checkRead, ref refsAdded);
+                    Socket.SocketListDangerousReleaseRefs(checkWrite, ref refsAdded);
+                    Socket.SocketListDangerousReleaseRefs(checkError, ref refsAdded);
+
+                    checkRead?.Clear();
+                    checkWrite?.Clear();
+                    checkError?.Clear();
+                }
+                else
+                {
+                    FilterPollList(checkRead, events, checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLHUP, ref refsAdded);
+                    FilterPollList(checkWrite, events, checkWriteInitialCount + checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLOUT, ref refsAdded);
+                    FilterPollList(checkError, events, checkErrorInitialCount + checkWriteInitialCount + checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLPRI, ref refsAdded);
+                }
+
+                return SocketError.Success;
             }
-            else
+            finally
             {
-                FilterPollList(checkRead, events, checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLHUP);
-                FilterPollList(checkWrite, events, checkWriteInitialCount + checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLOUT);
-                FilterPollList(checkError, events, checkErrorInitialCount + checkWriteInitialCount + checkReadInitialCount - 1, Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLPRI);
+                // This order matches with the AddToPollArray calls.
```

Comment:
It'd be worth calling out that the order must exactly match because if an error occurs part way through, we need to release only those we add ref'd.